### DISCUSSION TITLE
SSR performances issues on a website with only ~30 visits / day

### DISCUSSION THREAD

**Discussion by louisandred** [vercel/next.js#53922] [Help]
### Summary

Hello,

As the title suggest, I have some performances related issues.

**My nextjs app runs on a digital ocean droplet with the following specs** : 

- *2 CPU*
- *4 GB RAM*
- *50G Disk*

**Software side, I run on this droplet** : 

- *the nextjs app*
- *a nodejs server that acts as the api*
- *postgresql database with ~10 millions rows*
- *nginx reverse proxy*

I have ~30 visitors a day and my website is most of the time not responding, I'm trying to understand why.

I already looked at the api side, to see if the requests made too much times to complete, but the longest request takes ~300ms, and I added a simple cache (node-cache), to try to reduce some queries time.

The problem seems to be with the nextjs app.

Every pages are server side rendered (I need the SEO advantages).

I thought about switching to **static site generation**, but I have too many dynamic pages.
Plus I plan to add user authentication, so it will be a problem at some point because I would like to show users specifics data.

Then I thought too about switching to **incremental static regeneration**, but I don't really know for now how to handle user auth, that again, I plan to implement in the future.

But the point I was thinking is, I don't think **SSR** with nextjs is supposed to be a problem with only 30 visitors per day.

Which makes me think that I have a problem code wise with the app.
I don't know where to start.

Some help would be very much appreciated !

Thanks a lot 😄

PS : I also read this very interesting discussion https://github.com/vercel/next.js/discussions/10437, which gave me a lot of insight, but didn't really helped me with my problem.

**Comment by gerardnico** [comment]
Search on Google with the term `nextjs` and `profiling`.
Nextjs is just a node application therefore all articles on `NodeJs` and `profiling` will work also.

If this is too complicated, you can also add a bunch of debug log and see the output at this level.

**Reply by louisandred** [reply]
Hey, thank you very much for your answer !

What you said is a good start.

In the meantime, I searched for more information and it appears that running my `nodejs api` and my `nextjs app` both in cluster mode using `pm2` with the `instances` option set to `max` on a server with that kind of specs, is not a good idea.


Also, I found this issue https://github.com/vercel/next.js/issues/45508 with `jest-workers` that seemed to gave a bit more details on what was going on behind the scenes.
I followed the recommendations, switched to `next@canary` and killed the processes `jest-workers`, that reduced memory usage by a lot.


I do have a question though.
The following screenshot comes from my server after I installed `next@canary`. 

When I started the apps (`api` and `nextjs`), the memory consumption was at `~300M / 3.82G`.
I went on my website, loaded one or two pages, and the memory consumption rose to the current level, `~740M / 3.82G`.

Is it the expected behavior that the process `next-render-worker-pages` doesn't release the memory after the tasks have been completed ?

![nextjs-help](https://github.com/vercel/next.js/assets/67049116/ff77a2ca-834b-445d-b7b5-8ef6460f666c)


Thank you for your help !

**Reply by SouthLink** [reply]
Can you tell me what this process monitoring tool is, please

**Reply by louisandred** [reply]
Hey @SouthLink, look for the command line `htop` on linux :)
